### DISCUSSION TITLE
How to use MDC key value in logs ?

### DISCUSSION THREAD

**Discussion by rajagopal-mani** [quarkusio/quarkus#40479] [Q&A]
I am setting an MDC value in my code. I do not know how to map this MDC value into my logs.

Example RestEndpoint
```

@Path("/hello")
public class ExampleResource {

    Logger log = LoggerFactory.getLogger(ExampleResource.class);

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        MDC.put("correlation_id", UUID.randomUUID().toString());
        log.info("Sample log from Get Method");
        return "Hello from Quarkus REST";
    }
}
```
Logging configuration
```
quarkus.log.level=INFO
quarkus.log.min-level=DEBUG
quarkus.log.category."org.example".level=DEBUG

quarkus.log.console.json=true
quarkus.log.console.enable=true
quarkus.log.console.json.date-format=yyyy-MM-dd HH:mm:ss
quarkus.log.console.json.excluded-keys=sequence,loggerClassName,loggerName,level,threadName,threadId,ndc,hostName,processName,processId
quarkus.log.console.level=INFO
```
Dependency used:
```
<dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-logging-json</artifactId>
        </dependency>
```

Output now
`{"timestamp":"2024-05-06 21:47:44","message":"Sample log from Get Method","mdc":{"correlation_id":"03842e41-6e97-4bbc-bf23-9088872acb74"}}`

Expected Output
`{"timestamp":"2024-05-06 21:47:44","message":"Sample log from Get Method","correlation_id":"03842e41-6e97-4bbc-bf23-9088872acb74"}`

Can someone help on how to map mdc value for correlation id, into a separate field like here in expected output ?

Sample project can be cloned here https://github.com/angelfrend/quarkus-logging-demo.git

**Comment by cescoffier** [comment]
Try with: %X{<MDC property name>} in your log format. 

See https://quarkus.io/guides/logging

**Reply by rajagopal-mani** [reply]
I can change the console logging format and achieve this. Example below,
```
quarkus.log.console.format=%d{yyyy-MM-dd HH:mm:ss} %X{correlation_id} [%logger{36}] [%.36thread] %msg%n
quarkus.log.console.json=false
```
However I have no clue about how to achieve this for json log.
For example, there is an option to configure additional field for json logging as below,

```
quarkus.log.console.json.additional-field."correlation_id".value: %mdc{correlation_id}
quarkus.log.console.json.additional-field."correlation_id".type: string
```
I tried both %mdc{correlation_id} and %X{correlation_id}. Both did not produce anticipated result. I assume the additional fields can only handle static values ?

**Comment by dmlloyd** [comment]
At present I don't think the built-in JSON formatter supports MDC keys in the log. However there is a [Quarkiverse JSON extension](https://docs.quarkiverse.io/quarkus-logging-json/dev/index.html) which does support MDC configuration.

**Reply by rajagopal-mani** [reply]
I found below dependency which allows logback.xml to maintain all configurations.
```
 <groupId>io.quarkiverse.logging.logback</groupId>
      <artifactId>quarkus-logging-logback</artifactId>
      <version>1.1.2</version>
 </dependency>
```
Then I used regular logback.xml to manage logging configuration. I just couldn't figure how to forward the context parameters like application name from Quarkus to logback.xml. Otherwise it looks good.

Please let me know if incase I am doing something wrong.
